docs/schema/viz-1.0.schema.json

```diff
@@ -0,0 +1,295 @@
+          "description": "Whether icons that are close to each other should be collapsed together using the Leaflet.markerCluster plugin.",
+          "default": false,
+          "examples": [
+            true
+          ]
+        }
+      }
+    },
+    "VectorStyle": {
+      "type": "object",
+      "properties": {
+        "style": {
+          "type": "string",
+          "description": "The name of a preset in the VizWiz UI.",
+          "examples": [
+            "default"
+          ]
+        },
+        "color": {
+          "type": "string",
+          "pattern": "^#[a-fA-F0-9]{3,6}$",
+          "description": "Hex color to use as the stroke color on features. If a fill is applied, it will be this color at a semi-transparent opacity.",
+          "examples": [
+            "#0C2639"
+          ]
+        },
+        "hoverColor": {
+          "type": "string",
+          "pattern": "^#[a-fA-F0-9]{3,6}$",
+          "description": "Hex color to use as the stroke color on features when the mouse is hovered over them. If a fill is applied, it will be this color at a semi-transparent opacity.",
+          "examples": [
+            "#FB4D42"
+          ]
+        }
+      }
+    },
+    "LegendStyle": {
+      "type": "object",
+      "properties": {
+        "symbol": {
+          "type": "string",
+          "enum": [
+            "icon",
+            "line",
+            "polygon"
+          ],
+          "description": "* `icon`: Shows the `markerUrl` icon from `icons`.\n* `line`: Renders a line with the `color` from the `polygons` configuration.\n* `polygon`: Renders a box with the `color` from the `polygons` configuration, with a semitransparent fill."
+        },
+        "label": {
+          "type": "string",
+          "description": "The label to show in the legend for this layer.",
+          "examples": [
+            "Boston City Council Districts"
+          ]
+        }
+      }
+    },
+    "AddressSearch": {
+      "type": "object",
+      "properties": {
+        "title": {
+          "type": "string",
+          "description": "Title above the search box",
+          "examples": [
+            "Find an address"
+          ]
+        },
+        "placeholder": {
+          "type": "string",
+          "description": "Placeholder to go within the search box when it’s empty.",
+          "examples": [
+            "Search for your address..."
+          ]
+        },
+        "zoomToResult": {
+          "type": "boolean",
+          "description": "If true, the map will zoom in close to a search result after it’s found."
+        },
+        "autoPopupDataSourceUid": {
+          "anyOf": [
+            {
+              "type": "null"
+            },
+            {
+              "type": "string"
+            }
+          ],
+          "description": "If not `null`, will open the popup for the specified layer when the search result is chosen."
+        }
+      }
+    },
+    "ArcGisFeatureService": {
```

Comment:
My intention here is that other data sources would be their own types, since they’d likely be configured with different fields. The `data` field of `DataSource` has an `anyOf` that could accommodate more types as we make them.

Reply:
That's exactly why it was called `attributes`. It could be an object the declares what type it is, and then include whatever additional attributes required for that service.

It also makes it easier to instantiate an editor, since you can index the editors by "type" and pass the other attributes instead of having to search the object for various potential names of configuration keys.

This also doesn't address the redundancy of having a "type" in the object that doesn't need it, and how we handle the potential conflict of handling multiple of these data source definition objects when they are all named differently. As soon as we add a CSV type, for example, you could have `ArcGisFeatureService` along with a `CsvSource`, and which is correct?

It would mean major changes to the schema with every new data source type added or data source feature added. Instead we can leave it open with the only required attribute of the `attributes` object being `type` and the rest is optional.

Reply:
Here’s an example of how additional data source types would be added:

```
…
"definitions": {
  "DataSource": {
    "type": "object",
    "properties": {
      "data": {
        "anyOf": [
          { "$ref": "#/definitions/ArcGisFeatureServiceData" },
          { "$ref": "#/definitions/CsvData"}
        ]
      }
    }
  },
  "ArcGisFeatureServiceData": {
    "type": "object",
    "properties": {
      "type": {
        "type": "string",
        "enum": ["arcgis"]
      },
      "service": …,
      "layer": …
    }
  },
  "CsvData": {
    "type": "object",
    "properties": {
      "type": {
        "type": "string",
        "enum": ["csv"]
      },
      "url": …
    }
  }
}
…
```

The `type` property is a signal to what schema type is being used.

From an implementation standpoint I think this is equivalent to the current `attributes` object with a `type` property, it’s just that it defines what properties in `attributes` correspond to a given `type` value.

Both TypeScript and Flow can handle type checking these situations. (_I.e._ if you do a `switch` statement on the value of `type` it will know what properties are defined in each `case` branch.)

Reply:
In order to instantiate the editor, you will have to to a large `if...then` loop to see which data source type it is:

```
if (datasource.hasOwnProperty('CsvData')) {
    dataeditor = new CsvDataEditor(datasource.CsvData)
} else if (datasource.hasOwnProperty('ArcGisFeatureServiceData') {
    dataeditor = new ArcGisFeatureServiceData(datasource.ArcGisFeatureServiceData
}
```

Which is really inefficient, vs

```
dataeditor = new dataEditors[datasource.attributes.type](datasource.attributes)
```

What you are proposing requires much more maintenance to keep up and parse.

Reply:
Ah, ok. I can see where our disconnect is coming from. The JSON-Schema is a little tricky to read. If you pull down the branch, run `npm run docs`, and navigate to the schema documentation then this will probably be clearer.

We’re actually talking about the same thing.

In all cases the property is still `data` (the equivalent of your `attributes` above). "`ArcGisFeatureServiceData`" doesn’t actually appear as a property name in the data, it‘s just a type in the schema.

So, your shorter line of code definitely applies to this. The schema is just putting in the additional constraint that if the value of `data.type` is `"arcgis"` then `data` will have `service` and `layer` properties. In my previous example, if `data.type` is `"csv"` then `data` has a `url` property.

Reply:
Ah! That makes so much more sense.

Thanks for the patience in walking me through this!